src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2563 @@
+            this.sendData(stream_id, payload, close and !stream.waitForTrailers);
+        } else if (bun.String.tryFromJS(data_arg, globalObject)) |bun_str| {
+            var zig_str = bun_str.toUTF8(bun.default_allocator);
+            defer zig_str.deinit();
+            var payload = zig_str.slice();
+            this.sendData(stream_id, payload, close and !stream.waitForTrailers);
+        } else {
+            globalObject.throw("Expected data to be an ArrayBuffer or a string", .{});
+            return .zero;
+        }
+
+        if (close) {
+            if (stream.waitForTrailers) {
+                this.dispatch(.onWantTrailers, JSC.JSValue.jsNumber(stream.id));
+            }
+        }
+
+        return JSC.JSValue.jsBoolean(true);
+    }
+
+    fn getNextStreamID(this: *H2FrameParser) u32 {
+        var stream_id: u32 = this.lastStreamID;
+        if (stream_id % 2 == 0) {
+            stream_id += 1;
+        } else if (stream_id == 0) {
+            stream_id = 1;
+        } else {
+            stream_id += 2;
+        }
+
+        return stream_id;
+    }
+
+    pub fn request(this: *H2FrameParser, globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSValue {
+        JSC.markBinding(@src());
+        // we use PADDING_STRATEGY_NONE with is default
+        // TODO: PADDING_STRATEGY_MAX AND PADDING_STRATEGY_ALIGNED
+
+        const args_list = callframe.arguments(3);
+        if (args_list.len < 2) {
+            globalObject.throw("Expected headers and sensitiveHeaders arguments", .{});
+            return .zero;
+        }
+
+        const headers_arg = args_list.ptr[0];
+        const sensitive_arg = args_list.ptr[1];
+
+        if (!headers_arg.isObject()) {
+            globalObject.throw("Expected headers to be an object", .{});
+            return .zero;
+        }
+
+        if (!sensitive_arg.isObject()) {
+            globalObject.throw("Expected sensitiveHeaders to be an object", .{});
+            return .zero;
+        }
+
+        // max frame size will be always at least 16384
+        var buffer: [16384 - FrameHeader.byteSize - 5]u8 = undefined;
+        var header_buffer: [MAX_HPACK_HEADER_SIZE]u8 = undefined;
+        @memset(&buffer, 0);
+
+        var encoded_size: usize = 0;
+
+        var stream_id: u32 = this.getNextStreamID();
+        if (stream_id > MAX_STREAM_ID) {
+            return JSC.JSValue.jsNumber(-1);
+        }
+
+        const stream = this.handleReceivedStreamID(stream_id) orelse {
+            return JSC.JSValue.jsNumber(-1);
+        };
+
+        var headers_obj = headers_arg.asObjectRef();
+
+        // we iterate twice, because pseudo headers must be sent first, but can appear anywhere in the headers object
+        var iter = JSC.JSPropertyIterator(.{
+            .skip_empty_name = false,
+            .include_value = true,
+        }).init(globalObject, headers_obj);
+        defer iter.deinit();
+        for (0..2) |ignore_pseudo_headers| {
```

Comment:
if this loop throws, is there any cleanup necessary to do with the `stream`?

Reply:
Fixed in https://github.com/oven-sh/bun/pull/6761/commits/946a993adea0d7084e96012fd60337da5be5bc49